components/builder-api-client/src/builder.rs

```diff
@@ -878,22 +923,23 @@ impl BuilderAPIProvider for BuilderAPIClient {
         };
         debug!("Reading from {}", &pa.path.display());
 
-        let mut reader: Box<dyn Read> = if let Some(mut progress) = progress {
+        let reader: Box<dyn Read + Send> = if let Some(mut progress) = progress {
             progress.size(file_size);
             Box::new(TeeReader::new(file, progress))
         } else {
             Box::new(file)
         };
 
-        let result = self.add_authz(self.0.post_with_custom_url(&path, custom), token)
-                         .body(Body::SizedBody(&mut reader, file_size))
-                         .send();
+        let body = Body::sized(reader, file_size);
+        let resp = self.add_authz(self.0.post_with_custom_url(&path, custom), token)
+                       .body(body)
+                       .send()?;
+        debug!("Status: {:?}", resp.status());
 
-        match result {
-            Ok(Response { status: StatusCode::Created,
-                          .. }) => Ok(()),
-            Ok(response) => Err(err_from_response(response)),
-            Err(e) => Err(Error::from(e)),
+        if resp.status() != StatusCode::OK {
+            Err(err_from_response(resp))
+        } else {
+            Ok(())
```

Comment:
I left a similar comment elsewhere, and this is totally personal preference, but I think this reads better as:

```rust
if resp.status() == StatusCode::Ok {
    Ok(())
} else {
    Err(err_from_response(resp))
}
```

Similarly elsewhere.

Reply:
I went back and forth on this, and ended up keeping the negated checks first (as there were already many of those) and after a while it seemed to read a bit better that way to me. I'll look at it again...

Reply:
Thinking a bit more, I get why you might want to get the negated scenario out of the way first. If you know it's bad, just get rid of it now; otherwise move on.

I was messing around a bit, and a nice alternative might be a macro like this: 

```rust
macro_rules! fail_if_not {
    ($resp:expr, $status:expr) => {
        if $resp.status() != $status {
            return Err(err_from_response($resp));
        }
    };
}
```

This is nice because you can recast the above code like this:

```rust
fail_if_not!(resp, StatusCode::OK);
Ok(())
```
When the non-error case is more complex, the benefits are greater.

This lets you quickly get the errors out of the way, and lets you do so in a pretty descriptive and low-overhead way. You don't need to add an extra level of indentation to the code with an `if`/`else` block. Because it's a macro, you can only use this in functions that return a `Result`, but that's clearly not an issue here.

I played around with a function formulation, but then you have to concern yourself with return types and potentially reassigning the response, which can be done and is valid, but adds a bit of unnecessary ceremony and line noise in my opinion.

Reply:
And because it's a macro, you could also modify it to accept multiple status codes, a pattern I see in at least one other place in this module.

Reply:
Nice... a set of macros for these common patterns could be useful..

Reply:
Addressed by other fixes